cupyx/scipy/ndimage/morphology.py

```diff
@@ -1,13 +1,595 @@
+
+    .. seealso:: :func:`scipy.ndimage.iterate_structure`
+    """
+    if iterations < 2:
+        return structure.copy()
+    ni = iterations - 1
+    shape = [ii + ni * (ii - 1) for ii in structure.shape]
+    pos = [ni * (structure.shape[ii] // 2) for ii in range(len(shape))]
+    slc = tuple(
+        slice(pos[ii], pos[ii] + structure.shape[ii], None)
+        for ii in range(len(shape))
+    )
+    out = cupy.zeros(shape, bool)
+    out[slc] = structure != 0
+    out = binary_dilation(out, structure, iterations=ni)
+    if origin is None:
+        return out
+    else:
+        origin = _util._fix_sequence_arg(origin, structure.ndim, 'origin', int)
+        origin = [iterations * o for o in origin]
+        return out, origin
+
+
+def generate_binary_structure(rank, connectivity, *, on_cpu=False):
+    """Generate a binary structure for binary morphological operations.
+
+    Args:
+        rank(int): Number of dimensions of the array to which the structuring
+            element will be applied, as returned by `np.ndim`.
+        connectivity(int): `connectivity` determines which elements of the
+            output array belong to the structure, i.e., are considered as
+            neighbors of the central element. Elements up to a squared distance
+            of `connectivity` from the center are considered neighbors.
+            `connectivity` may range from 1 (no diagonal elements are
+            neighbors) to `rank` (all elements are neighbors).
+        on_cpu(bool, optional): If True, return a NumPy array rather than
+            transferring the structure to the GPU. (Default value = False)
+
+    Returns:
+        cupy.ndarray: Structuring element which may be used for binary
+            morphological operations, with `rank` dimensions and all dimensions
+            equal to 3.
+
+    .. seealso:: :func:`scipy.ndimage.generate_binary_structure`
+    """
+    if connectivity < 1:
+        connectivity = 1
+    if rank < 1:
+        if on_cpu:
+            return numpy.asarray(True, dtype=bool)
+        else:
+            return cupy.asarray(True, dtype=bool)
+    output = numpy.fabs(numpy.indices([3] * rank) - 1)
+    output = numpy.add.reduce(output, 0)
+    output = output <= connectivity
+    if on_cpu:
+        return output
+    return cupy.asarray(output)
+
+
+def _binary_erosion(input, structure, iterations, mask, output, border_value,
+                    origin, invert, brute_force=True):
+    try:
+        iterations = operator.index(iterations)
+    except TypeError:
+        raise TypeError('iterations parameter should be an integer')
+
+    if not input.flags.c_contiguous:
+        # TODO: grlee77: is C-contiguity required?
+        input = cupy.ascontiguousarray(input)
+    if input.dtype.kind == 'c':
+        raise TypeError('Complex type not supported')
+    if structure is None:
+        structure = generate_binary_structure(input.ndim, 1)
+    else:
+        structure = cupy.asarray(structure, dtype=bool)
```

Comment:
Please use `astype` is the given `structure` is guaranteed to be `cupy.ndarray`.